### ISSUE TITLE
buildRustPackage() requires change of cargoDeps checksum when version changes

### ISSUE THREAD

**Issue by limeytexan** [NixOS/nixpkgs#112763]
**Describe the bug**
When using `buildRustPackage()` a change of `version` causes the resulting `cargoDeps` path to change which then requires that the expression's `cargoHash` / `cargoSha256` checksum be similarly updated. This in turn prevents the use of differing `version` strings with the one Nix expression.

**To Reproduce**
1. clone https://github.com/limeytexan/flox-rust-testing
2. from within the clone, invoke: `nix-build`
3. then invoke it a second time overriding the `version` string: `nix-build --argstr version "1.1.1"`

**Expected behavior**
In the above example I would have expected to be able to use the same Nix expression for the same package built with different version strings, but instead find that we must update the Nix expression `cargoDeps` checksum in order to build it when versions change.

**Screenshots**
```
[brantley@clubsoda:~/src/flox-rust-testing]$ cat default.nix
{ version ? "1.0.0" }:

with import <nixpkgs> { };

rustPlatform.buildRustPackage rec {
  pname = "flox-rust-testing";
  inherit version;
  src = lib.cleanSource ./.;
  cargoSha256 = "0h62pbl5ybyxlk587vala3jp0k0hlpxg6v7x2988w6ga9b04g40g";
}

[brantley@clubsoda:~/src/flox-rust-testing]$ nix-build
/nix/store/35m20jw97hwkbxavpbm5ynsfg39pgb2v-flox-rust-testing-1.0.0

[brantley@clubsoda:~/src/flox-rust-testing]$ nix-build --argstr version "1.1.1"
these derivations will be built:
  /nix/store/pcpdaxffc32c6z1nydvpy3zq0kmkadwl-flox-rust-testing-1.1.1-vendor.tar.gz.drv
  /nix/store/yvykisjidzwzq843nh2h5jsc63cs6lan-flox-rust-testing-1.1.1.drv
building '/nix/store/pcpdaxffc32c6z1nydvpy3zq0kmkadwl-flox-rust-testing-1.1.1-vendor.tar.gz.drv'...
unpacking sources
unpacking source archive /nix/store/x0zl5kzch507ab5l9db9v1wg7l17m9pf-source
source root is source
patching sources
building
To use vendored sources, add this to your .cargo/config.toml for this project:

installing
hash mismatch in fixed-output derivation '/nix/store/n242bpzr8w7xpjvbx7c4awlirppzqkx3-flox-rust-testing-1.1.1-vendor.tar.gz':
  wanted: sha256:0h62pbl5ybyxlk587vala3jp0k0hlpxg6v7x2988w6ga9b04g40g
  got:    sha256:1bq33r6k6nq0rf4dwv89hcbmv5gpq1azryf1lh2ykp9f2yfcd7cd
cannot build derivation '/nix/store/yvykisjidzwzq843nh2h5jsc63cs6lan-flox-rust-testing-1.1.1.drv': 1 dependencies couldn't be built
error: build of '/nix/store/yvykisjidzwzq843nh2h5jsc63cs6lan-flox-rust-testing-1.1.1.drv' failed

[brantley@clubsoda:~/src/flox-rust-testing]$ 
```

**Additional context**
This problem can be fixed by updating the arguments to `fetchCargoTarball()` to use `args.pname` instead of `name`. I will submit a PR with this change momentarily.

@Infinisil also reports that the suggested fix is in keeping with how other source fetchers use the name source for the derivation to avoid similar problems.

This issue was originally raised as [a topic on the flox discourse forum](https://discourse.floxdev.com/t/mismatch-cargosha256-hash-between-manual-build-and-flox-ci/118/7).

**Notify maintainers**

It's not exactly straightforward to identify maintainers for this builder but a few guesses:
* from `rustc.meta.maintainers`: @madjar @cstrahan @globin @havvy
* from `git blame` of the `fetchCargoTarball` invocation: @alyssais @shlevy @danieldk @Ekleog @Mic92 
* generally interested: @grahamc @Infinisil

**Comment by danieldk** [comment]
The current behavior has benefits and downsides:

* Downside 1: you need to update `cargoSha256` even if the vendored dependencies don't change (as you describe).
* Downside 2: it goes against fixed-output derivations/content addressing, the vendored dependencies should be addressable by the content hash, not by the `name` of the derivation that uses them.
* Benefit: you cannot accidentally forget to update `cargoSha256`, since the output changes with the version bump.

**Comment by veprbl** [comment]
Duplicate of #85703

**Comment by Mic92** [comment]
> The current behavior has benefits and downsides:
> 
>     * Downside 1: you need to update `cargoSha256` even if the vendored dependencies don't change (as you describe).
> 
>     * Downside 2: it goes against fixed-output derivations/content addressing, the vendored dependencies should be addressable by the content hash, not by the `name` of the derivation that uses them.
> 
>     * Benefit: you cannot accidentally forget to update `cargoSha256`, since the output changes with the version bump.

Do we not also catch this with https://github.com/NixOS/nixpkgs/blob/65aeed59c52c572eccdf9237ce45d9fb0fd18b91/pkgs/build-support/rust/default.nix#L172 ?

**Comment by danieldk** [comment]
> Do we not also catch this with

Indeed, we would.

**Comment by limeytexan** [comment]
Addressed by #113176 (thanks!), closing.
